### ISSUE TITLE
Feature request: Add a new `uv create` command to scaffold new projects (like `hatch new` or `yarn create`)

### ISSUE THREAD

**Issue by hellmrf** [astral-sh/uv#9754]
## Problem Statement

Starting a new project often involves repetitive tasks like configuring project structures, setting up tools, and applying team-specific templates. While these steps are common across projects, they remain manual and error-prone within the `uv` ecosystem. Developers need:

- Predefined templates with standard structures and settings.
- Scaffolding for frameworks (e.g., Django, FastAPI, or React).
- Standardized configurations for CI/CD, linters, or documentation (e.g., `.gitignore`, README, LICENSE).

Existing tools like [`hatch`](https://hatch.pypa.io/latest/cli/reference/#hatch-new) and [`cookiecutter`](https://cookiecutter.readthedocs.io/) address these issues but lack integration with `uv`, leading to fragmented workflows for developers who rely on `uv` as their project manager. `npm`-world tools like [`create-react-app`](https://create-react-app.dev/docs/getting-started#creating-an-app), available via [`npm init`](https://docs.npmjs.com/cli/v8/commands/npm-init) or [`yarn create`](https://classic.yarnpkg.com/blog/2017/05/12/introducing-yarn/), are also great and popular options, sadly specific to Node.js ecosystem.

## Proposed Solution: `uv create`

Introduce a `uv create` subcommand to streamline project bootstrapping. This command would allow users to scaffold projects using predefined or custom templates, inspired by tools like `hatch new`, `create-react-app`, and `cookiecutter`.

#### Using Built-in Templates

```bash
$ uv create django my_project
$ cd my_project
$ uv run start
```

#### Using Custom Templates from GitHub

```bash
$ uv create myuser/mytemplate my_project
$ cd my_project
$ uv run start
```

#### Alternative Syntax: `uv init --from`

For alignment with `npm init` and avoiding redundancy with `uv init`, an alternative could implementing this as a `uv init` feature:

```bash
$ mkdir my_project && cd my_project
$ uv init --from django
```

For custom templates:
```bash
$ uv init --from myuser/mytemplate
```

## Benefits of `uv create`

- **Seamless integration:** Unified workflow within `uv`.
- **Flexibility:** Support for both built-in and user-defined templates.
- **Productivity:** Automates repetitive tasks, reducing setup time and errors.

## Contribution

If the team approves, I’m happy to contribute to the implementation with guidance on project structure and integration points.

**Comment by zanieb** [comment]
Thanks for the thorough overview. How does this differ from https://github.com/astral-sh/uv/issues/4759?

**Comment by hellmrf** [comment]
First of all, I confess I did an extensive search in this repo for related issues (I even used Copilot!) but didn’t find the related issue, so I apologize for any oversight. That said, I believe there are significant differences between this proposal and #4759, so I will share my thoughts on why this merits consideration.

While #4759 proposes **integrating external tools** like Cookiecutter to `uv init`, I'm focusing on **standardizing workflows** with a **fully native solution** (thinking ultimately in `uv` becoming the Python's `cargo`).

### Cookiecutter?

[Cookiecutter](https://github.com/cookiecutter/cookiecutter), while popular and language-agnostic, faces challenges such as a slow development activity ([< 50 commits on 2024](https://github.com/cookiecutter/cookiecutter/commits/main)) and more than [200 opened issues](https://github.com/cookiecutter/cookiecutter/issues), including problems like failing to run `mypy` on templates ([cookiecutter/cookiecutter#2116](https://github.com/cookiecutter/cookiecutter/issues/2116)). 

Relying on such external tools could undermine `uv`'s reputation for speed and reliability, particularly as a Rust-based project.

A native `uv create` command, as outlined here, would eliminate these dependencies, ensuring stability, performance, and alignment with `uv`'s core strengths. This approach not only avoids potential bottlenecks but also positions `uv` as a robust, standalone solution capable of setting new standards for project management in Python and beyond.

### Learn from other people's mistakes and successes

While Cookiecutter is a well-known tool, **it hasn't become the standard for project initialization**, leaving workflows fragmented across different teams and ecosystems. 

In contrast, tools like [`create-react-app`](https://create-react-app.dev/docs/getting-started#creating-an-app), [`npm init`](https://docs.npmjs.com/cli/v8/commands/npm-init), and [`yarn create`](https://classic.yarnpkg.com/en/docs/cli/create/) have established themselves as **widely adopted standards** in the Node.js/Deno ecosystems (e.g. [React Native's Get Started Docs](https://reactnative.dev/docs/environment-setup#start-a-new-react-native-project-with-expo)). 

These tools, being the standard in the industry, provide stability, uniformity, and a consistent developer experience, making it easier for new users to onboard and adopt best practices seamlessly. 

### Conclusion 

This way, implementing a native `uv create` command would bring similar benefits to the `uv` ecosystem, offering an intuitive and standardized approach to project scaffolding that developers and teams can rely on.

**Comment by zanieb** [comment]
Thanks for explaining! I'm going to close this as a duplicate, since we can track the functionality in the other issue — but I'll link to this context.

**Comment by zanieb** [comment]
Actually sorry to flip flop, I think I'm okay tracking this separately if you're suggesting we define our _own_ template tooling — that's the case, yes?

**Comment by hellmrf** [comment]
@zanieb yes, that's correct: I'm proposing the **development of a new, `uv`-specific, template tooling** to standardize workflows within the `uv` ecosystem. For that, I'm arguing that, while cookiecutter users can `uv tool run cookiecutter`, the `uv create` approach should ensure tight integration, enhance performance, and provide a consistent user experience, aligning with `uv`'s goals of speed and reliability.

**Comment by baggiponte** [comment]
I'd love for uv to develop its own template tooling! On the other hand, there are quite a lot of templates lying around and it would be a waste to not be able to support them.

A built-in template engine would be really dope and could add features that are lacking in engines such as cookiecutter - chief of all, the ability to update the current project to a more recent version of a template.

**Comment by kevinhav** [comment]
I would also like to see this supported by `uv`. In my opinion project setup is one of the more cumbersome tasks, and it seems to fit well in `uv` from a workflow perspective.

**Comment by astrojuanlu** [comment]
> chief of all, the ability to update the current project to a more recent version of a template.

Have a look at https://copier.readthedocs.io !

**Comment by josecannete** [comment]
I think this could be a great addition to uv!

**Comment by lucafaggianelli** [comment]
I second this, it's a great addition to uv!

For starting small, as a PoC, why don't we just support a hit repository or folder copy without any templating support? I think this would already solve many use cases.

**Comment by hellmrf** [comment]
> For starting small, as a PoC, why don't we just support a hit repository or folder copy without any templating support? I think this would already solve many use cases.

I actually agree. The template repository could just have a setup script and do the templating however they want.

**Comment by pasunboneleve** [comment]
An alternative would be to direct template users to a language-agnostic scaffolding engine and instead providing a standard `uv` configuration for it. That's what I did, BTW: [baker-uv-template](https://github.com/dmvianna/baker-uv-template).

**Comment by aretrace** [comment]
Having a `uv` equivalent to `npm create` would be **awesome**.

**Comment by kamikaze** [comment]
please

**Comment by julie777** [comment]
As a user, I want to use a project tool that supports full project creation and not just a tiny part of it. UV recognizes this value with `init --lib` and `init --app` but doesn't really take it very far. 

**Installation Based Templating Support**

Use UV tool install to install a templateing tool such as cookiecutter or copier, and because UV installed the tool it knows about it. When a recognized templating tool is installed by UV then UV should automatically add support to the init command.
```
uv init --cc <template name> ...              # use cookiecutter with its config to find templates
uv init --cp <template name> ...              # use copier with its config to find templates by name
```

**Built In Support for Templates**

UV could decide to always include one of the templating applications as part of its install and provide a few standard templates for projects. This could replace --app and --lib.

In addition, I find that the UV init does not do all the init that Hatch does, such as LICENSE, test directory, etc. Instead of hard-coding project types into UV just supporting an existing templating mechanism would be easier and keep the UV codebase cleaner.

**Comment by pasunboneleve** [comment]
@julie777 that is not a big ask, and you can make it happen outside of `uv`. 

I created my personal `uv create` wrapper with `baker`. It achieves exactly the goal you are aiming for. Check it out: [baker-uv-template](https://github.com/pasunboneleve/baker-uv-template)

**Comment by jhessin** [comment]
All I want to do is inject this in all my new projects pyproject.toml

```toml

[tool.ruff]
preview = true

[tool.ruff.lint]
select = ["E", "F", "I"]

[tool.ruff.format]
quote-style = 'single'
docstring-code-format = true
```

What is the simplest way to do that?
